content/blog/2019-12-18-bril-osr/index.md

```diff
@@ -0,0 +1,373 @@
+be deconstructed.
+
+Implementing returning values from Bril functions was considerably easier, as
+they can just be placed in the return register (`rax`), and then forwarded back
+through `handle_call` to the caller.
+
+[Godbolt]: https://godbolt.org
+
+#### Combining the JIT with the Interpreter
+
+To improve the performance of the JIT, we combined it with the interpreter. We
+extended the interpreter to perform basic profiling. It maintains a counter for
+each function, keeping track of how many times that function has run. Then, once
+that counter reaches a fixed value, the program will then compile the function
+instead of interpreting it. After that, every time the function is called, the
+compiled version can be simply run.
+
+### On-stack replacement
+
+To implement on-stack replacement, we extend the profiling information from the
+previous section to keep track of the number of times each basic block in a
+function is executed. If a function is being interpreted and one of these
+counters reaches a certain fixed value, OSR will be initiated immediately. Code
+generation will then procede as before, but with a few OSR-specific additions.
+The basic structure of a function compiled via OSR is as follows:
+
+```
+osr_start:
+    # prologue
+    # move variable values to stack
+    jmp osr_first_inst
+
+regular_start:
+    # prologue
+    # ...
+osr_first_inst:
+    # ...
+```
+
+As such, the assembly has additional code at `osr_start` which will only execute
+this one time for completing the on-stack replacement. Future calls to this
+function will ignore all of that and start from the `regular_start`. In this new
+section of the assembly, we move the current values of all of the variables to
+their appropriate positions in the stack, and then jump to whatever instruction
+the interpreter was about to execute.
+
+## Evaluation
+
+To evaluate Bril-OSR, we aimed to first verify its correctness using the
+existing [brili](https://github.com/sampsyo/bril) interpreter as a benchmark,
+and then its performance on a series of reasonably large-scale Bril tests.
+
+All tests were run on an Intel Core i7-7700HQ CPU @ 2.80GHz with 16GB of RAM,
+and using Ubuntu on WSL.
+
+## Correctness Tests
+
+First, we generated multiple programs that covered the breadth of Bril opcodes
+from a configurable Python script. We ran the correctness suite on both the
+interpreter and just-in-time code generation engine, and ensured that the
+runtime effects matched those of Brili.
+
+**Note**: Inspired by Alexa and Gregory's foray into
+[Hypothesis](https://hypothesis.works/), we attempted to write property-based
+tests for Bril-OSR using Rust's
+[quickcheck](https://docs.rs/quickcheck/0.9.0/quickcheck/) crate. Unfortunately,
+generating structures representing Bril programs was difficult, so we abandoned
+this project. Still, this would be a cool extension to Bril-OSR's correctness
+test suite.
+
+## Performance Tests (Ablation Study)
```

Comment:
FWIW, outside of machine learning, this kind of thing is more commonly called a "sensitivity study."